internal/lib/sandbox/builder.go

```diff
@@ -123,37 +123,112 @@ type Builder interface {
 
 	// SetCreatedAt sets the created at time.
 	SetCreatedAt(createdAt time.Time)
+
+	// Validate validates the sandbox.
+	Validate() error
+}
+
+type sandboxValidations uint64
+
+const (
+	validationSetID sandboxValidations = 1 << iota
+	validationSetCRISandbox
+	validationSetCreatedAt
+	validationSetName
+	validationSetLogDir
+	validationSetShmPath
+	validationSetNamespace
+	validationSetKubeName
+	validationSetProcessLabel
+	validationSetMountLabel
+	validationSetCgroupParent
+	validationSetPrivileged
+	validationSetRuntimeHandler
+	validationSetResolvPath
+	validationSetHostname
+	validationSetPortMappings
+	validationSetHostNetwork
+	validationSetUsernsMode
+	validationSetPodLinuxOverhead
+	validationSetPodLinuxResources
+)
+
+// setValidation sets the given validation flag.
+func (v *sandboxValidations) setValidation(flag sandboxValidations) {
+	*v |= flag
+}
+
+// isSet returns true if the given validation flag is set.
+func (v *sandboxValidations) isSet(flag sandboxValidations) bool {
+	return *v&flag != 0
 }
 
 // sandboxBuilder is the hidden default type behind the Sandbox interface.
 type sandboxBuilder struct {
-	config     *types.PodSandboxConfig
-	infra      container.Container
-	sandboxRef *Sandbox
+	config      *types.PodSandboxConfig
+	infra       container.Container
+	sandboxRef  *Sandbox
+	validations sandboxValidations
 }
 
 // NewBuilder creates a new, empty Sandbox instance.
 func NewBuilder() Builder {
 	return &sandboxBuilder{
-		config:     nil,
-		sandboxRef: new(Sandbox),
+		config:      nil,
+		sandboxRef:  new(Sandbox),
+		validations: 0,
 	}
 }
 
 // GetSandbox gets the sandbox and deletes the config and sandbox.
 // TODO: Add validations before returning the sandbox.
 func (b *sandboxBuilder) GetSandbox() (*Sandbox, error) {
-	if b.sandboxRef.criSandbox == nil {
-		return nil, errors.New("cri-o sandbox not initialized")
+	if err := b.Validate(); err != nil {
+		return nil, err
 	}
 
 	sandboxRef := b.sandboxRef
 	b.config = nil
 	b.sandboxRef = nil
+	b.validations = 0
 
 	return sandboxRef, nil
 }
 
+// Validate validates the sandbox.
+func (b *sandboxBuilder) Validate() error {
```

Comment:
one more nit: we could short circut this by storing a const of all of the validations and then directly comparing that value to the validations . Then we could loop through the map to report what wasn't done

Reply:
done